File: env-shield/node/envshield.js
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

let envShieldCache = null;
let masterKeyCache = null;

function getMasterKey() {
    if (masterKeyCache) return masterKeyCache;
    
    // 1. Check ENV_SHIELD_KEY in environment
    if (process.env.ENV_SHIELD_KEY) {
        masterKeyCache = Buffer.from(process.env.ENV_SHIELD_KEY, 'hex');
        return masterKeyCache;
    }
    
    // 2. Check local env-shield.key file
    const keyPath = path.resolve(process.cwd(), 'env-shield.key');
    if (fs.existsSync(keyPath)) {
        masterKeyCache = Buffer.from(fs.readFileSync(keyPath, 'utf8').trim(), 'hex');
        return masterKeyCache;
    }
    
    throw new Error("Master key not found. Set ENV_SHIELD_KEY or provide env-shield.key file.");
}

function loadEncryptedData() {
    if (envShieldCache) return envShieldCache;
    const encPath = path.resolve(process.cwd(), '.env.enc');
    if (!fs.existsSync(encPath)) {
        return {};
    }
    envShieldCache = JSON.parse(fs.readFileSync(encPath, 'utf8'));
    return envShieldCache;
}

function decryptValue(encryptedObj) {
    const key = getMasterKey();
    const iv = Buffer.from(encryptedObj.iv, 'hex');
    const authTag = Buffer.from(encryptedObj.auth_tag, 'hex');
    const ciphertext = Buffer.from(encryptedObj.ciphertext, 'base64');
    
    // Verify HMAC
    const hmacKey = crypto.createHash('sha256').update(key).digest();
    const hmac = crypto.createHmac('sha256', hmacKey);
    hmac.update(Buffer.concat([iv, ciphertext]));
    const computedTag = hmac.digest();
    
    if (!crypto.timingSafeEqual(authTag, computedTag)) {
        throw new Error("Authentication failed! Data has been tampered with.");
    }
    
    const decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
    let decrypted = decipher.update(ciphertext, undefined, 'utf8');
    decrypted += decipher.final('utf8');
    return decrypted;
}

function init() {
    // Return a Proxy wrapping process.env
    return new Proxy(process.env, {
        get(target, prop) {
            // Priority 1: native unencrypted process.env values
            if (prop in target) {
                return target[prop];
            }
            
            // Priority 2: Check JIT encrypted values
            const data = loadEncryptedData();
            if (prop in data) {
                return decryptValue(data[prop]);
            }
            
            return undefined;
        }
    });
}

// Automatically replace process.env on require
process.env = init();

module.exports = process.env;

File: env-shield/python/envshield.py
import os
import json
import base64
import hmac
import hashlib
import subprocess

def _get_master_key():
    # 1. Check ENV_SHIELD_KEY in environment
    if 'ENV_SHIELD_KEY' in os.environ:
        return bytes.fromhex(os.environ['ENV_SHIELD_KEY'])
        
    # 2. Check Colab userdata
    try:
        from google.colab import userdata
        key = userdata.get('ENV_SHIELD_KEY')
        if key:
            return bytes.fromhex(key)
    except ImportError:
        pass
        
    # 3. Check local env-shield.key file
    key_path = os.path.join(os.getcwd(), 'env-shield.key')
    if os.path.exists(key_path):
        with open(key_path, 'r') as f:
            return bytes.fromhex(f.read().strip())
            
    raise RuntimeError("Master key not found. Set ENV_SHIELD_KEY or provide env-shield.key")

def _load_enc_data():
    enc_path = os.path.join(os.getcwd(), '.env.enc')
    if not os.path.exists(enc_path):
        return {}
    with open(enc_path, 'r') as f:
        return json.load(f)

def _decrypt_value(encrypted_obj):
    key = _get_master_key()
    iv = bytes.fromhex(encrypted_obj['iv'])
    auth_tag = bytes.fromhex(encrypted_obj['auth_tag'])
    ciphertext = base64.b64decode(encrypted_obj['ciphertext'])
    
    # Verify HMAC (Auth Tag)
    hmac_key = hashlib.sha256(key).digest()
    computed_tag = hmac.new(hmac_key, iv + ciphertext, hashlib.sha256).digest()
    
    if not hmac.compare_digest(auth_tag, computed_tag):
        raise RuntimeError("Authentication failed! Data has been tampered with.")
        
    # Decrypt via openssl
    cmd = ['openssl', 'enc', '-d', '-aes-256-cbc', '-K', key.hex(), '-iv', iv.hex()]
    proc = subprocess.run(cmd, input=ciphertext, capture_output=True)
    if proc.returncode != 0:
        raise RuntimeError(f"OpenSSL decryption failed: {proc.stderr.decode('utf-8')}")
        
    return proc.stdout.decode('utf-8')

# We patch os.environ.__class__.__getitem__ and .get dynamically
_original_getitem = os.environ.__class__.__getitem__
_original_get = os.environ.__class__.get
_enc_cache = None

def _envshield_getitem(self, key):
    try:
        return _original_getitem(self, key)
    except KeyError:
        global _enc_cache
        if _enc_cache is None:
            _enc_cache = _load_enc_data()
            
        if key in _enc_cache:
            return _decrypt_value(_enc_cache[key])
        raise

def _envshield_get(self, key, default=None):
    try:
        return _envshield_getitem(self, key)
    except KeyError:
        return default

os.environ.__class__.__getitem__ = _envshield_getitem
os.environ.__class__.get = _envshield_get
